test/command/suite/select/function/snippet/invalid/too_long_key.actual

```diff
@@ -0,0 +1,37 @@
+[
+  [
+    [
+      -22,
+      0.0,
+      0.0
+    ],
+    "[hash][add][(temporary)] too long key"
+  ],
+  [
+    [
+      [
+        1
+      ],
+      [
+        [
+          "snippet",
+          null
+        ]
+      ],
+      [
+        "[hash][add][(temporary)] too long key"
+      ]
+    ]
+  ]
+]
+#|e| [hash][add][(temporary)] too long key
```

Comment:
Why is this occurred? What is the problem of this case?

Reply:
I thought it was the intended behavior without giving it much thought, but on second thought, it seems a bug.

proc_snippet.c
```
    grn_proc_func_generate_cache_key(ctx, "snippet", args + 1, nargs - 1, &cache_key);
    ...
    snip_ptr = grn_expr_get_or_add_var(ctx,
                                       expression,
                                       GRN_TEXT_VALUE(&cache_key),
                                       GRN_TEXT_LEN(&cache_key));
```

`grn_proc_func_generate_cache_key` adds all arguments as string to `cache_key` and `GRN_TEXT_LEN` returns the length  including these arguments.
`GRN_TEXT_LEN(&cache_key))`  is greater than the max hash key size (4096) on this case.

I am not confident whether passing `GRN_TEXT_LEN(&cache_key)` to `grn_expr_get_or_add_var` is correct or not.
Do you have any good idea to fix it?
Anyway, I will create a new issue or PR.

Reply:
> I am not confident whether passing GRN_TEXT_LEN(&cache_key) to grn_expr_get_or_add_var is correct or not.

I guess this intends to cache function results including arguments, but in other cases, simply passing a name and its length, what is the difference?

proc_highlight.c
```
    highlighter_ptr =
      grn_expr_get_or_add_var(ctx, expression,
                              GRN_FUNC_HIGHLIGHT_HTML_CACHE_NAME,
                              strlen(GRN_FUNC_HIGHLIGHT_HTML_CACHE_NAME));
```

Reply:
Ah, I understand.
We should not use cache when the cache key is too large:

```diff
diff --git a/lib/proc/proc_snippet.c b/lib/proc/proc_snippet.c
index d101d502d..60f61b8aa 100644
--- a/lib/proc/proc_snippet.c
+++ b/lib/proc/proc_snippet.c
@@ -115,10 +115,12 @@ func_snippet(grn_ctx *ctx, int nargs, grn_obj **args, grn_user_data *user_data)
   grn_obj *snip_ptr = NULL;
   grn_obj *expression;
   grn_proc_get_info(ctx, user_data, NULL, NULL, &expression);
-  snip_ptr = grn_expr_get_var(ctx,
+  if (GRN_TEXT_LEN(&cache_key) <= GRN_TABLE_MAX_KEY_SIZE) {
+    snip_ptr = grn_expr_get_var(ctx,
                               expression,
                               GRN_TEXT_VALUE(&cache_key),
                               GRN_TEXT_LEN(&cache_key));
+  }
   if (snip_ptr) {
     snip = GRN_PTR_VALUE(snip_ptr);
   } else {
```

Cache is just for performance. We should not use cache if we don't provide normal feature with cache.

We can use a static cache key (`GRN_FUNC_HIGHLIGHT_HTML_CACHE_NAME`) for `highlight_html` because it doesn't have any arguments. It means that all `highlight_html` call should return the same result. So we can cache it with the static cache key.

Reply:
Thank you.
I have modified to check `GRN_TEXT_LEN(&cache_key)`.